extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/data/input/kafkainput/KafkaInputReader.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.java.util.common.parsers.CloseableIterator;
+import org.apache.druid.java.util.common.parsers.ParseException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+public class KafkaInputReader implements InputEntityReader
+{
+  private static final Logger log = new Logger(KafkaInputReader.class);
+  private static final String DEFAULT_KEY_STRING = "key";
+  private static final String DEFAULT_TIMESTAMP_STRING = "timestamp";
+
+  private final InputRowSchema inputRowSchema;
+  private final KafkaRecordEntity record;
+  private final KafkaHeaderReader headerParser;
+  private final InputEntityReader keyParser;
+  private final InputEntityReader valueParser;
+  private final String keyLabelPrefix;
+  private final String recordTimestampLabelPrefix;
+
+  public KafkaInputReader(
+      InputRowSchema inputRowSchema,
+      KafkaRecordEntity record,
+      KafkaHeaderReader headerParser,
+      InputEntityReader keyParser,
+      InputEntityReader valueParser,
+      String keyLabelPrefix,
+      String recordTimestampLabelPrefix
+  )
+  {
+    this.inputRowSchema = inputRowSchema;
+    this.record = record;
+    this.headerParser = headerParser;
+    this.keyParser = keyParser;
+    this.valueParser = valueParser;
+    this.keyLabelPrefix = keyLabelPrefix;
+    this.recordTimestampLabelPrefix = recordTimestampLabelPrefix;
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read() throws IOException
+  {
+    Map<String, Object> mergeList = new HashMap<>(headerParser.read());
+    // Add kafka record timestamp to the mergelist
+    mergeList.put(recordTimestampLabelPrefix + DEFAULT_TIMESTAMP_STRING, record.getRecord().timestamp());
+
+    // Return type for the key parser should be of type MapBasedInputRow
+    // Parsers returning other types are not compatible currently.
+    if (keyParser != null) {
+      try (CloseableIterator<InputRow> keyIterator = keyParser.read()) {
+        // Key currently only takes the first row and ignores the rest.
+        if (keyIterator.hasNext()) {
+          MapBasedInputRow keyRow = (MapBasedInputRow) keyIterator.next();
+          mergeList.put(
+              keyLabelPrefix + DEFAULT_KEY_STRING,
+              keyRow.getEvent().entrySet().stream().findFirst().get().getValue()
+          );
+        }
+      }
+      catch (Exception e) {
+        if (e instanceof IOException) {
+          log.error(e, "Encountered IOException during key parsing.");
+          throw (IOException) e;
+        } else if (e instanceof ParseException) {
+          log.error(e, "Encountered key parsing exception.");
+        } else {
+          log.error(e, "Encountered exception during key parsing.");
+          throw e;
+        }
+      }
+    }
+
+    List<InputRow> rows = new ArrayList<>();
```

Comment:
materializing all the rows into a list will increase burden on the garbage collector. The reason to have the CloseableIterator in the first place is so we can stream back the rows instead of materializing them all before returning them. We should see if we can avoid materializing anything other than the fields added from the key and the header, and stream back the rest.